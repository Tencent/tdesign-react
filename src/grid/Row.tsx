import React, { useEffect, useState, forwardRef, createContext } from 'react';
import classNames from 'classnames';
import isObject from 'lodash/isObject';
import useConfig from '../hooks/useConfig';
import { StyledProps } from '../common';
import { TdRowProps } from './type';
import { canUseDocument, getCssVarsValue } from '../_util/dom';
import { rowDefaultProps } from './defaultProps';

/**
 * Row 组件支持的属性。
 */
export interface RowProps extends TdRowProps, StyledProps {
  /**
   * 默认子元素内容
   */
  children: React.ReactNode;
}

const calcSize = (width: number) => {
  const smWidth = getCssVarsValue('--td-screen-sm') || 768;
  const mdWidth = getCssVarsValue('--td-screen-md') || 992;
  const lgWidth = getCssVarsValue('--td-screen-lg') || 1200;
  const xlWidth = getCssVarsValue('--td-screen-xl') || 1400;
  const xxlWidth = getCssVarsValue('--td-screen-xxl') || 1880;

  let size = 'xs';
  if (width >= xxlWidth) {
    size = 'xxl';
  } else if (width >= xlWidth) {
    size = 'xl';
  } else if (width >= lgWidth) {
    size = 'lg';
  } else if (width >= mdWidth) {
    size = 'md';
  } else if (width >= smWidth) {
    size = 'sm';
  } else {
    size = 'xs';
  }

  return size;
};

const calcRowStyle = (gutter: TdRowProps['gutter'], currentSize: string): object => {
  const rowStyle = {};
  if (typeof gutter === 'number') {
    Object.assign(rowStyle, {
      marginLeft: `${gutter / -2}px`,
      marginRight: `${gutter / -2}px`,
    });
  } else if (Array.isArray(gutter) && gutter.length) {
    if (typeof gutter[0] === 'number') {
      Object.assign(rowStyle, {
        marginLeft: `${gutter[0] / -2}px`,
        marginRight: `${gutter[0] / -2}px`,
      });
    }
    if (typeof gutter[1] === 'number') {
      Object.assign(rowStyle, { rowGap: `${gutter[1]}px` });
    }

    if (isObject(gutter[0]) && gutter[0][currentSize] !== undefined) {
      Object.assign(rowStyle, {
        marginLeft: `${gutter[0][currentSize] / -2}px`,
        marginRight: `${gutter[0][currentSize] / -2}px`,
      });
    }
    if (isObject(gutter[1]) && gutter[1][currentSize] !== undefined) {
      Object.assign(rowStyle, { rowGap: `${gutter[1][currentSize]}px` });
    }
  } else if (isObject(gutter) && gutter[currentSize]) {
    if (Array.isArray(gutter[currentSize]) && gutter[currentSize].length) {
      Object.assign(rowStyle, {
        marginLeft: `${gutter[currentSize][0] / -2}px`,
        marginRight: `${gutter[currentSize][0] / -2}px`,
      });
      Object.assign(rowStyle, { rowGap: `${gutter[currentSize][1]}px` });
    } else {
      Object.assign(rowStyle, {
        marginLeft: `${gutter[currentSize] / -2}px`,
        marginRight: `${gutter[currentSize] / -2}px`,
      });
    }
  }
  return rowStyle;
};

export const RowContext = createContext({ gutter: undefined, size: undefined });

const Row = forwardRef((props: RowProps, ref) => {
  const { align, gutter, justify, tag, style: propStyle, className, children, ...otherRowProps } = props;

  const [size, setSize] = useState(canUseDocument ? calcSize(window.innerWidth) : 'md');

  const updateSize = () => {
    const currentSize = calcSize(window.innerWidth);
    if (currentSize !== size) {
      setSize(size);
    }
  };

  const { classPrefix } = useConfig();
  const rowClassNames = classNames(`${classPrefix}-row`, className, {
    [`${classPrefix}-row--${justify}`]: true,
    [`${classPrefix}-row--${align}`]: true,
  });
  const rowStyle = {
    ...calcRowStyle(gutter, size),
    ...propStyle,
  };

  useEffect(() => {
    window.addEventListener('resize', updateSize);
    return () => {
      window.removeEventListener('resize', updateSize);
    };
  });

  return React.createElement(
    tag,
    {
      ref,
      className: rowClassNames,
      style: rowStyle,
      ...otherRowProps,
    },
    <RowContext.Provider value={{ gutter, size }}>{children}</RowContext.Provider>,
  );
});

Row.displayName = 'Row';
Row.defaultProps = rowDefaultProps;

export default Row;
