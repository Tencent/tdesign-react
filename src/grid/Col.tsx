import React, { useContext, forwardRef } from 'react';
import classNames from 'classnames';
import isObject from 'lodash/isObject';
import useConfig from '../hooks/useConfig';
import { StyledProps } from '../common';
import { TdColProps, TdRowProps } from './type';
import { colDefaultProps } from './defaultProps';
import { RowContext } from './Row';

type FlexType = number | 'none' | 'auto' | string;

export interface ColProps extends TdColProps, StyledProps {
  /**
   * 文本内容
   */
  children?: React.ReactNode;
}

const calcColPadding = (gutter: TdRowProps['gutter'], currentSize: string) => {
  const paddingObj = {};
  if (typeof gutter === 'number') {
    Object.assign(paddingObj, {
      paddingLeft: `${gutter / 2}px`,
      paddingRight: `${gutter / 2}px`,
    });
  } else if (Array.isArray(gutter) && gutter.length) {
    if (typeof gutter[0] === 'number') {
      Object.assign(paddingObj, {
        paddingLeft: `${gutter[0] / 2}px`,
        paddingRight: `${gutter[0] / 2}px`,
      });
    }

    if (isObject(gutter[0]) && gutter[0][currentSize]) {
      Object.assign(paddingObj, {
        paddingLeft: `${gutter[0][currentSize] / 2}px`,
        paddingRight: `${gutter[0][currentSize] / 2}px`,
      });
    }
  } else if (isObject(gutter) && gutter[currentSize]) {
    Object.assign(paddingObj, {
      paddingLeft: `${gutter[currentSize] / 2}px`,
      paddingRight: `${gutter[currentSize] / 2}px`,
    });
  }
  return paddingObj;
};

const parseFlex = (flex: FlexType) => {
  if (typeof flex === 'number') {
    return `${flex} ${flex} auto`;
  }

  if (/^\d+(\.\d+)?(px|r?em|%)$/.test(flex)) {
    return `0 0 ${flex}`;
  }

  return flex;
};

const Col = forwardRef((props: ColProps | any, ref) => {
  const { flex, offset, order, pull, push, span, tag, className, children, style: propStyle, ...otherColProps } = props;
  const { gutter: rowGutter, size: rowSize } = useContext(RowContext);

  const { classPrefix } = useConfig();
  const allSizes = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'];
  const sizeClasses = allSizes.reduce((acc, currSize) => {
    const sizeProp = props[currSize];
    let sizeObj: any = {};
    if (typeof sizeProp === 'number') {
      sizeObj.span = sizeProp;
    } else if (isObject(sizeProp)) {
      sizeObj = sizeProp || {};
    }

    return {
      ...acc,
      [`${classPrefix}-col-${currSize}-${sizeObj.span}`]: sizeObj.span !== undefined,
      [`${classPrefix}-col-${currSize}-order-${sizeObj.order}`]: parseInt(sizeObj.order, 10) >= 0,
      [`${classPrefix}-col-${currSize}-offset-${sizeObj.offset}`]: parseInt(sizeObj.offset, 10) >= 0,
      [`${classPrefix}-col-${currSize}-push-${sizeObj.push}`]: parseInt(sizeObj.push, 10) >= 0,
      [`${classPrefix}-col-${currSize}-pull-${sizeObj.pull}`]: parseInt(sizeObj.pull, 10) >= 0,
    };
  }, {});

  const colClassNames = classNames(
    `${classPrefix}-col`,
    className,
    {
      [`${classPrefix}-col-${span}`]: span !== undefined,
      [`${classPrefix}-col-offset-${offset}`]: parseInt(offset, 10) >= 0,
      [`${classPrefix}-col-pull-${pull}`]: parseInt(pull, 10) >= 0,
      [`${classPrefix}-col-push-${push}`]: parseInt(push, 10) >= 0,
      [`${classPrefix}-col-order-${order}`]: parseInt(order, 10) >= 0,
    },
    sizeClasses,
  );

  const colStyle = {
    ...calcColPadding(rowGutter, rowSize),
    ...propStyle,
  };
  flex && (colStyle.flex = parseFlex(flex));

  return React.createElement(
    tag,
    {
      ref,
      className: colClassNames,
      style: colStyle,
      ...otherColProps,
    },
    children,
  );
});

Col.displayName = 'Col';
Col.defaultProps = colDefaultProps;

export default Col;
