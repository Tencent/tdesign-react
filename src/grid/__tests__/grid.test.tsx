import React from 'react';
import { render } from '@test/utils';
import Row from '../Row';
import Col from '../Col';

describe('Grid 组件测试', () => {
  test('Grid ', async () => {
    window.innerWidth = 1500;
    const { container } = render(
      <Row gutter={12}>
        <Col></Col>
      </Row>,
    );
    // 校验默认className
    const defaultClass = ['t-row'];
    expect(container.firstChild).toHaveClass(...defaultClass);
  });

  test('Row gutter', async () => {
    // 校验默认className
    const defaultClass = ['t-row'];
    window.innerWidth = 1300;

    // number
    const { container: container1 } = render(
      <Row gutter={12}>
        <Col span={1}></Col>
      </Row>,
    );

    // Array<number>
    const { container: container2 } = render(
      <Row gutter={[6, 6]}>
        <Col span={2}></Col>
      </Row>,
    );

    // Array<GutterObject>
    const { container: container3 } = render(
      <Row
        gutter={[
          { xs: 4, sm: 8, md: 12 },
          { xs: 4, sm: 8, md: 12 },
        ]}
      >
        <Col span={3}></Col>
      </Row>,
    );

    // Object
    const { container: container4 } = render(
      <Row gutter={{ xs: 4, sm: 8, md: 12 }}>
        <Col span={4}></Col>
      </Row>,
    );

    const col1 = container1.querySelector('.t-col-1');
    expect(col1).not.toBeNull();
    const col2 = container2.querySelector('.t-col-2');
    expect(col2).not.toBeNull();
    const col3 = container3.querySelector('.t-col-3');
    expect(col3).not.toBeNull();
    const col4 = container4.querySelector('.t-col-4');
    expect(col4).not.toBeNull();

    expect(container1.firstChild).toHaveClass(...defaultClass);
    expect(container2.firstChild).toHaveClass(...defaultClass);
    expect(container3.firstChild).toHaveClass(...defaultClass);
    expect(container4.firstChild).toHaveClass(...defaultClass);
  });

  test('Col flex', async () => {
    window.innerWidth = 800;
    // number
    const { container: container1 } = render(
      <Row>
        <Col flex={2}>
          <div>2 / 5</div>
        </Col>
        <Col flex={3}>
          <div>3 / 5</div>
        </Col>
      </Row>,
    );

    // string
    const { container: container2 } = render(
      <Row gutter={12}>
        <Col flex={'100px'}>
          <div>2 / 5</div>
        </Col>
        <Col flex={'100px'}>
          <div>3 / 5</div>
        </Col>
      </Row>,
    );

    // auto
    const { container: container3 } = render(
      <Row gutter={12}>
        <Col flex={'auto'}>
          <div>2 / 5</div>
        </Col>
        <Col flex={'auto'}>
          <div>3 / 5</div>
        </Col>
      </Row>,
    );
    expect(container1.firstChild.firstChild).toHaveStyle({ flex: '2 2 auto' });
    expect(container2.firstChild.firstChild).toHaveStyle({ flex: '0 0 100px' });
    expect(container3.firstChild.firstChild).toHaveStyle({ flexBasis: 'auto' });
  });
});
