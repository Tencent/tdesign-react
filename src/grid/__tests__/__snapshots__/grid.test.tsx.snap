// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="t-col t-col-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        3
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="t-col t-col-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="t-col t-col-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        4
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-6"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        6
      </div>
    </div>
    <div
      class="t-col t-col-6"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        6
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-12"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        12
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`flex.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 2 2 auto;"
    >
      <div>
        2 / 5
      </div>
    </div>
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 3 3 auto;"
    >
      <div>
        3 / 5
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 0 0 100px;"
    >
      <div>
        100px
      </div>
    </div>
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex-basis: auto;"
    >
      <div>
        Fill Rest
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 1 1 200px;"
    >
      <div>
        1 1 200px
      </div>
    </div>
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 0 1 300px;"
    >
      <div>
        0 1 300px
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex: 0 0 auto;"
    >
      <div
        style="padding: 0px 16px;"
      >
        none
      </div>
    </div>
    <div
      class="t-col"
      style="padding-left: 0px; padding-right: 0px; flex-basis: auto;"
    >
      <div>
        auto with no-wrap
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`gutter.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: -8px; margin-right: -8px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: -12px; margin-right: -12px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 12px; padding-right: 12px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 12px; padding-right: 12px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 12px; padding-right: 12px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 12px; padding-right: 12px;"
    >
      <div>
        col-3
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: -8px; margin-right: -8px; row-gap: 24px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div>
        col-3
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`halign.jsx 1`] = `
<DocumentFragment>
  <p>
    align left
  </p>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
  </div>
  <p>
    align center
  </p>
  <div
    class="t-row t-row--center t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
  </div>
  <p>
    align right
  </p>
  <div
    class="t-row t-row--end t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
  </div>
  <p>
    space-between
  </p>
  <div
    class="t-row t-row--space-between t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
  </div>
  <p>
    space-around
  </p>
  <div
    class="t-row t-row--space-around t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
    <div
      class="t-col t-col-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`offset.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-4
      </div>
    </div>
    <div
      class="t-col t-col-4 t-col-offset-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-4
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3 t-col-offset-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-3 col-offset-3
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-offset-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-3 col-offset-3
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-6 t-col-offset-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-6 col-offset-2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`order.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    通过 \`order\` 来改变元素的排序。
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3 t-col-order-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        1 col-order-4
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-order-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        2 col-order-3
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-order-2"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        3 col-order-2
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-order-1"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        4 col-order-1
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3 t-col-xs-order-1 t-col-sm-order-2 t-col-md-order-3 t-col-lg-order-4"
      lg="[object Object]"
      md="[object Object]"
      sm="[object Object]"
      style="padding-left: 0px; padding-right: 0px;"
      xs="[object Object]"
    >
      <div>
        1 col-order-responsive
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-xs-order-2 t-col-sm-order-1 t-col-md-order-4 t-col-lg-order-3"
      lg="[object Object]"
      md="[object Object]"
      sm="[object Object]"
      style="padding-left: 0px; padding-right: 0px;"
      xs="[object Object]"
    >
      <div>
        2 col-order-responsive
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-xs-order-3 t-col-sm-order-4 t-col-md-order-2 t-col-lg-order-1"
      lg="[object Object]"
      md="[object Object]"
      sm="[object Object]"
      style="padding-left: 0px; padding-right: 0px;"
      xs="[object Object]"
    >
      <div>
        3 col-order-responsive
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-xs-order-4 t-col-sm-order-3 t-col-md-order-1 t-col-lg-order-2"
      lg="[object Object]"
      md="[object Object]"
      sm="[object Object]"
      style="padding-left: 0px; padding-right: 0px;"
      xs="[object Object]"
    >
      <div>
        4 col-order-responsive
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`responsive.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-xs-2 t-col-sm-4 t-col-md-6 t-col-lg-8 t-col-xl-10"
      lg="8"
      md="6"
      sm="4"
      style="padding-left: 0px; padding-right: 0px;"
      xl="10"
      xs="2"
    >
      <div>
        Col
      </div>
    </div>
    <div
      class="t-col t-col-xs-10 t-col-sm-8 t-col-md-6 t-col-lg-4 t-col-xl-2"
      lg="4"
      md="6"
      sm="8"
      style="padding-left: 0px; padding-right: 0px;"
      xl="2"
      xs="10"
    >
      <div>
        Col
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`sort.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    通过 \`pull\` \`push\` 进行排序
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-9 t-col-push-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-9 col-push-3
      </div>
    </div>
    <div
      class="t-col t-col-3 t-col-pull-9"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-3 col-pull-9
      </div>
    </div>
  </div>
  <div
    class="t-row t-row--start t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-8 t-col-push-4"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-8 col-push-4
      </div>
    </div>
    <div
      class="t-col t-col-4 t-col-pull-8"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div>
        col-4 col-pull-8
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`valign.jsx 1`] = `
<DocumentFragment>
  <p>
    align top
  </p>
  <div
    class="t-row t-row--center t-row--top"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
  </div>
  <p>
    Align Middle
  </p>
  <div
    class="t-row t-row--space-around t-row--middle"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
  </div>
  <p>
    Align Bottom
  </p>
  <div
    class="t-row t-row--space-between t-row--bottom"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 80px;"
      >
        col-3
      </div>
    </div>
    <div
      class="t-col t-col-3"
      style="padding-left: 0px; padding-right: 0px;"
    >
      <div
        style="height: 40px;"
      >
        col-3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
